gluejobutils/scd.py

```diff
@@ -0,0 +1,209 @@
+    """
+    Updates the data that current exists at table_db_base_path with new_df.
+
+    new_df should have the exact same columns and meta as table_db_base_path (therefore record start and end dates should have been
+    applied to new_df.
+
+    table at table_db_base_path is read in and joined to new_df on update_by_cols. Any rows from table_db_base_path 
+    that do not match to new_df are written to a partition called dea_record_update_type (set to old). These are records that are unchanged. 
+
+    The rest of table_db_base_path is appended to new_df as these rows also exist in new_df so we need to recalculate SCD2.
+
+    The SCD2 (specifically dea_record_start_date and dea_record_end_date) are recalculated for the appended df. SCD2 are recalculated
+    using the update_by_cols variable. This data is then written the same tmp directory but with the partition dea_record_update_type set to new.
+
+    If no file exists in table_db_base_path then this function writes new_df straight to the tmp folder with partition dea_record_update_type set to new.
+
+    update_latest_records_only means that only latest records are called from current table. Set to false if you wish to back fill old update.
+    """
+    if not isinstance(update_by_cols, list) :
+        raise TypeError('update_by_cols must be a list of column names')
+
+    bucket, key = s3_path_to_bucket_key(table_db_base_path)
+    table_base_path, table_name = os.path.split(remove_slash(key))
+
+    table_db_path = add_slash(os.path.join(table_db_base_path, partition_path))
+    tmp_table_db_path = os.path.join(table_base_path, table_name + '_tmp', partition_path)
+    tmp_table_db_path_old_partition = add_slash(os.path.join('s3://', bucket, tmp_table_db_path, 'dea_record_update_type=old'))
+    tmp_table_db_path_new_partition = add_slash(os.path.join('s3://', bucket, tmp_table_db_path, 'dea_record_update_type=new'))
+
+    if folder_contains_only_files_with_extension(table_db_path) :
+        new_keys = new_df.select(*update_by_cols)
+        new_keys.createOrReplaceTempView('update_keys')
+        spark.read.parquet(table_db_path).cache().createOrReplaceTempView('current_table')
+
+        ### NB That his is optimised to only add latest records. Will not work when trying to add an older set of records
+        update_primary_keys_select = list_to_sql_select(update_by_cols, table_alias = 'update_keys')
+        update_concat_keys = 'CONCAT({})'.format(update_primary_keys_select)
+        db_primary_keys_select = list_to_sql_select(update_by_cols, table_alias = 'current_table')
+        db_concat_keys = 'CONCAT({})'.format(db_primary_keys_select)
+
+        where_statement = "WHERE current_table.dea_record_end_datetime=to_timestamp('{}', '{}')".format(static_record_end_datetime, standard_datetime_format) if update_latest_records_only else ""
+
```

Comment:
Probably add a comment here like:
`# Create a table containing only the records all the records in the current table, with a flag to indicate whether there's a corresponding key in the new records`